### PR TITLE
trie: more node iterator improvements

### PR DESCRIPTION
**Pull request by fjl**
Split out of #14589 to get review earlier.

The changes:

* NodeIterator interface gains `LeafKey` method which returns the key of the current (leaf) node.
* nodeIterator now handles missing nodes gracefully. Callers can recover from `MissingNodeError`
   by inserting the node into the database before calling `Next` again.
* nodeIterator no longer returns hash nodes and always adds the terminator symbol to `Path`.
  This makes `Path` unique.
* `MissingNodeError` contains hex-encoded Path of the missing node. This is useful for the light client.

### CODE REVIEW

File: trie/encoding.go [hunk 1]
```diff
@@ -75,13 +75,16 @@ func keybytesToHex(str []byte) []byte {
...
 // hexToKeybytes turns hex nibbles into key bytes.
-// This can only be used for keys of even length.
+// This is only useful for keys of even length.
 func hexToKeybytes(hex []byte) []byte {
 	if hasTerm(hex) {
 		hex = hex[:len(hex)-1]
 	}
 	if len(hex)&1 != 0 {
-		panic("can't convert hex key of odd length")
+		// Odd length. This can happen if someone iterates a Trie
```
Comment (Arachnid) [inline trie/encoding.go]: What's the purpose of this change?

File: trie/iterator.go [hunk 2]
```diff
@@ -89,69 +95,79 @@ type nodeIteratorState struct {
...
 	if trie.Hash() == emptyState {
 		return new(nodeIterator)
 	}
 	it := &nodeIterator{trie: trie}
-	it.seek(start)
+	it.err = it.seek(start)
 	return it
 }
 
-// Hash returns the hash of the current node
```
Comment (Arachnid) [inline trie/iterator.go]: Why delete these comments?

File: trie/iterator.go [hunk 3]
```diff
@@ -160,29 +176,34 @@ func (it *nodeIterator) Error() error {
...
+	if seek, ok := it.err.(seekError); ok {
+		if it.err = it.seek(seek.key); it.err != nil {
+			return false
+		}
+	}
+	// Otherwise step forward with the iterator and report any errors.
 	state, parentIndex, path, err := it.peek(descend)
-	if err != nil {
-		it.err = err
+	if it.err = err; it.err != nil {
```
Comment (Arachnid) [inline trie/iterator.go]: I don't think it makes sense to have the first statement here part of the if clause.

File: trie/encoding.go (hunk 1 above)
Comment (fjl) [inline trie/encoding.go]: It's explained in the commit message:

> This also removes the the panic for odd length because it is too subtle
> to be useful. In a follow-up commit, using LeafKey in a core/state test
> revealed that it iterates tries rooted in an intermediate node. This can
> produce odd-length leaf keys and it's hard to debug. Removing the panic
> removes risk of a rare production crash later.

File: trie/iterator.go (hunk 2 above)
Comment (fjl) [inline trie/iterator.go]: They're not visible in GoDoc and other implementations of NodeIterator don't have comments either.

File: trie/iterator.go (hunk 3 above)
Comment (fjl) [inline trie/iterator.go]: hmm, I guess this is a style question that we can 🚲 🏠 over.

File: trie/encoding.go (hunk 1 above)
Comment (Arachnid) [inline trie/encoding.go]: Isn't a 'rare production crash' the correct behaviour if we're iterating partial tries?

File: trie/iterator.go (hunk 3 above)
Comment (Arachnid) [inline trie/iterator.go]: I don't think it's bikeshedding. Would you agree this is bad style?
```
if it.err = err; err != nil {
```
they're functionally equivalent, but the latter demonstrates more clearly that the first part of the if statement is only being used for its side-effect.

Comment (fjl) [inline trie/iterator.go]: The first part of the if is always used for its side effect because it is a statement.

Comment (Arachnid) [inline trie/iterator.go]: Yes, but it's generally used to generate something; not just an assignment.

File: trie/encoding.go (hunk 1 above)
Comment (Arachnid) [inline trie/encoding.go]: I still don't understand what the compelling reason to return nil here is, though. Is there any situation in which this would be expected?

**Comment by fjl** [comment]
squashed
